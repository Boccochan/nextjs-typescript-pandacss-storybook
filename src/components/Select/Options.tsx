type OptionProps = {
  /**
   * Hidden this option on dropdown menu.
   */
  hidden?: boolean;

  /**
   * The selected attribute.
   */
  isSelected?: boolean;

  /**
   * The id is set as 'key' of option tag.
   */
  id: string;

  /**
   * The value is shown at dropdown menu.
   */
  value: string;
};

type OptionsProps = {
  /**
   * If true, the component is disabled and show 'Loading..'.
   */
  loading?: boolean;

  /**
   * The options of the select.
   */
  options: OptionProps[];
};

export const Options = (props: OptionsProps) => {
  return props.options.map((option) => (
    <option
      key={option.id}
      hidden={option.hidden}
      value={option.value}
      selected={option.isSelected}
    >
      {option.value}
    </option>
  ));
};
