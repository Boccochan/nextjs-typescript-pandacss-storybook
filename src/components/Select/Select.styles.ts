import { css, cva } from "#/styled-system/css";

const wrapper = css({
  position: "relative",
  width: "fit-content",
  _after: {
    position: "absolute",
    content: '""',
    width: "10px",
    height: "10px",
    top: "50%",
    right: "10px",
    transform: "translateY(-50%)",
    bg: "blue",
  },
});

const select = cva({
  base: {
    borderColor: "body.border",
    borderWidth: "1px",
    borderRadius: "md",
    bg: "body.bg",
    color: "body.text",
    appearance: "none",
    padding: "1px 2rem",
  },
  variants: {
    size: {
      sm: {
        px: "3",
        rounded: "sm",
        fontSize: "sm",
        minWidth: "3rem",
        height: "2rem",
      },
      md: {
        px: "4",
        rounded: "md",
        fontSize: "md",
        minWidth: "4rem",
        height: "2.4rem",
      },
      lg: {
        px: "5",
        rounded: "lg",
        fontSize: "lg",
        minWidth: "5rem",
        height: "2.8rem",
      },
      xl: {
        px: "6",
        rounded: "xl",
        fontSize: "xl",
        minWidth: "6rem",
        height: "3.2rem",
      },
    },
  },
  defaultVariants: {
    size: "md",
  },
});

export const styles = {
  select,
  wrapper,
};
