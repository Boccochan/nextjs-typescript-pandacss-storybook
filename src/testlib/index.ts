export * from "./render";
export * from "./translator";
