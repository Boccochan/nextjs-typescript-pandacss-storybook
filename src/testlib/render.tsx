import { render as rtlRender } from "@testing-library/react";
import { NextIntlClientProvider } from "next-intl";
import React from "react";

import messageEn from "../../messages/en.json";
import messageJa from "../../messages/ja.json";

type RenderTuple = Parameters<typeof rtlRender>;

const messages = {
  en: messageEn,
  ja: messageJa,
};

export const render = (
  ui: RenderTuple[0],
  options?: RenderTuple[1],
  local?: "en" | "ja",
) =>
  rtlRender(ui, {
    wrapper: ({ children }) => (
      <NextIntlClientProvider
        locale={local ?? "en"}
        messages={local ? messages[local] : messageEn}
      >
        {children}
      </NextIntlClientProvider>
    ),
    ...options,
  });
