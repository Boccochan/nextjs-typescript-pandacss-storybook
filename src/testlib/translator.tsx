import { createTranslator } from "next-intl";

import { locales } from "../i18n";

export const getTranslator = async (locale: string) => {
  if (!locales.includes(locale)) throw Error("Unkown locale");
  const messages = (await import(`../../messages/${locale}.json`)).default;

  return createTranslator({ locale, messages });
};
