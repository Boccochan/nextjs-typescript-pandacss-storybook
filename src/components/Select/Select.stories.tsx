import type { Meta, StoryObj } from "@storybook/react";
import type { ComponentProps } from "react";

import { css } from "#/styled-system/css";

import { Select } from "./Select";

type SelectProps = ComponentProps<typeof Select>;

const DisplayBox = (props: SelectProps) => {
  return (
    <div
      className={css({
        width: "500px",
        display: "flex",
        alignItems: "center",
        justifyContent: "center",
        height: "100px",
        borderColor: "body.border",
        borderWidth: "1px",
        borderRadius: "md",
        position: "relative",
        _before: {
          position: "absolute",
          content: '"Parent component"',
          top: "-3.5",
          left: "10px",
          bg: "body.bg",
          padding: "0 4px",
          color: "body.text",
        },
      })}
    >
      <Select {...props} />
    </div>
  );
};

const meta: Meta<typeof Select> = {
  title: "components/Select",
  component: Select,
  tags: ["autodocs"],
  parameters: {
    componentSubtitle:
      "The Select component should be used when there are 4 or more options. Otherwise, consider radio button.",
    layout: "centered",
    docs: {
      controls: { exclude: ["disabled", "onClick"] },
    },
  },
  argTypes: {
    size: {
      control: "select",
      options: ["xs", "sm", "md", "lg", "xl"],
    },
    width: {
      control: "select",
      options: ["xs", "sm", "md", "lg", "xl", "full", "auto"],
    },
  },
};

export default meta;
type Story = StoryObj<typeof Select>;

export const Normal: Story = {
  args: {
    disabled: false,
    size: "md",
    width: "auto",
    options: [
      { hidden: true, id: "Select", value: "Select" },
      { id: "Apple", value: "Apple" },
      {
        id: "International Consolidated Airlines Group SA",
        value: "International Consolidated Airlines Group SA",
      },
      { id: "Amazon", value: "Amazon" },
      { id: "Meta", value: "Meta" },
      { id: "Google", value: "Google" },
    ],
  },
  render: DisplayBox,
};

export const Empty: Story = {
  args: {
    size: "md",
    width: "auto",
    disabled: false,
    options: [],
  },
  render: DisplayBox,
};

export const Selected: Story = {
  args: {
    size: "md",
    width: "auto",
    disabled: false,
    options: [
      { hidden: true, id: "Select", value: "Select" },
      { id: "Apple", value: "Apple" },
      {
        id: "International Consolidated Airlines Group SA",
        value: "International Consolidated Airlines Group SA",
      },
      { id: "Amazon", value: "Amazon", isSelected: true },
      { id: "Meta", value: "Meta" },
      { id: "Google", value: "Google" },
    ],
  },
  render: DisplayBox,
};

export const Loading: Story = {
  args: {
    size: "md",
    width: "auto",
    disabled: false,
    loading: true,
    options: [
      { hidden: true, id: "Select", value: "Select" },
      { id: "Apple", value: "Apple" },
      {
        id: "International Consolidated Airlines Group SA",
        value: "International Consolidated Airlines Group SA",
      },
      { id: "Amazon", value: "Amazon", isSelected: true },
      { id: "Meta", value: "Meta" },
      { id: "Google", value: "Google" },
    ],
  },
  render: DisplayBox,
};
