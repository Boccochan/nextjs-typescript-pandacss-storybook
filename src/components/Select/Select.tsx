import React from "react";

import { styles } from "./Select.styles";

type Option = {
  hidden?: boolean;
  value: string;
};

type SelectProps = React.JSX.IntrinsicElements["select"] & {
  options: Option[];
};

export const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
  ({ options, disabled, ...rest }, ref) => (
    <div className={styles.wrapper({ disabled })}>
      <select
        ref={ref}
        {...rest}
        className={styles.select()}
        disabled={disabled}
      >
        {options.map((option, index) => (
          <option key={index} hidden={option.hidden}>
            {option.value}
          </option>
        ))}
      </select>
    </div>
  ),
);

Select.displayName = "Select";
