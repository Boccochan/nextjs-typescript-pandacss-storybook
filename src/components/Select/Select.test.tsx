import React from "react";

import { render } from "@/testlib";

import { Select } from "./Select";

describe("Snapshot Select", () => {
  it("With size=xs, options=[]", async () => {
    const select = render(<Select size="xs" options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With size=sm, options=[]", async () => {
    const select = render(<Select size="sm" options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With size=md, options=[]", async () => {
    const select = render(<Select size="md" options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With size=lg, options=[]", async () => {
    const select = render(<Select size="lg" options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With size=xl, options=[]", async () => {
    const select = render(<Select size="xl" options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With options=[]", async () => {
    const select = render(<Select options={[]} />);

    expect(select).toMatchSnapshot();
  });

  it("With options=['Apple', 'Microsoft']", async () => {
    const select = render(
      <Select
        options={[
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With options=['Apple', 'Microsoft'] disabled", async () => {
    const select = render(
      <Select
        options={[
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
        disabled
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With loading", async () => {
    const select = render(<Select options={[]} loading />);

    expect(select).toMatchSnapshot();
  });

  it("With options=['Apple', 'Microsoft'] selected Microsoft", async () => {
    const select = render(
      <Select
        options={[
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft", isSelected: true },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With options=['Select', 'Apple', 'Microsoft'] select option", async () => {
    const select = render(
      <Select
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft", isSelected: true },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='xs' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="xs"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='sm' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="sm"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='md' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="md"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='lg' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="lg"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='xl' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="xl"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='auto' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="auto"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });

  it("With width='full' options=['Select', 'Apple', 'Microsoft'] ", async () => {
    const select = render(
      <Select
        width="full"
        options={[
          { id: "select", value: "Select", hidden: true },
          { id: "apple", value: "Apple" },
          { id: "microsoft", value: "Microsoft" },
        ]}
      />,
    );

    expect(select).toMatchSnapshot();
  });
});
