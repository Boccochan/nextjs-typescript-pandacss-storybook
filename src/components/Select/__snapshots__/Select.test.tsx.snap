// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Select With size=xs, options=[] 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pos_relative w_fit-content after:pos_absolute after:content_"" after:top_50% after:right_10px after:transform_translateY(-50%) after:w_0 after:h_0 after:border-l_6px_solid_transparent after:border-r_6px_solid_transparent after:border-t_icon after:border-tw_8px after:border-top-style_solid after:pointer-events_none opacity_0.4"
      >
        <select
          class="border_body.border border-w_1px rounded_xs bg_body.bg text_body.text appearance_none p_1px_2rem_1px_1rem disabled:opacity_0.7 disabled:pointer-events_none px_2 fs_xs min-w_10rem h_1.6rem w_auto"
          disabled=""
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pos_relative w_fit-content after:pos_absolute after:content_"" after:top_50% after:right_10px after:transform_translateY(-50%) after:w_0 after:h_0 after:border-l_6px_solid_transparent after:border-r_6px_solid_transparent after:border-t_icon after:border-tw_8px after:border-top-style_solid after:pointer-events_none opacity_0.4"
    >
      <select
        class="border_body.border border-w_1px rounded_xs bg_body.bg text_body.text appearance_none p_1px_2rem_1px_1rem disabled:opacity_0.7 disabled:pointer-events_none px_2 fs_xs min-w_10rem h_1.6rem w_auto"
        disabled=""
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
